### DISCUSSION TITLE
Mixing params, prefix and wildcard

### DISCUSSION THREAD

**Discussion by hardillb** [fastify/fastify#4394] [Q&A] [answered]
Hi,

I'm trying to mix route params, prefixes and wildcards and getting strange results

e.g. if I do the following:

```
await app.registery(require('./foo.js`, { prefix: `/foo` })
```

and `foo.js` includes:

```
const fp = require('fastify-plugin')

module.exports = fp(async function (app, opts, done) {

  app.get('/:teamId/:projectid'/*', async (request, reply) => {
    const teamId = request.params.teamId
    cont projectId = request.params.projectId
    const path = request.params['*']

    console.log(teamId)
    console.log(projectId)
    ...
  })

done()
})
```

When I run `curl http://localhost:3000/foo/team/project/random/path` I would expect:

- teamId = team
- projectId = project
- path = random/path

But I'm getting

- teamId = foo
- projectId = team
- path = project/random/path

If I remove the prefix and specify the full path when defining the `get` handler it all appears to work properly.

Any Idea how to fix this or is it a bug I should raise?

Or do I have to use a regexp pattern as hinted at in https://github.com/fastify/fastify/discussions/3347 ?

I'm running NodeJS 16.x and Fastify 4.9.2

**Comment by climba03003** [comment, marked as answer]
Are you wrapping `foo.js` with `fastify-plugin`?
If yes, then it is expected behavior.

**Reply by hardillb** [reply]
Yes, `foo.js` is a fastify-plugin (updated the question to make that clear)

But I don't get why it's expected behaviour, it's counter intuitive and **very** none obvious and I still don't know how to get behaviour I've described (without dropping the use of `prefix` which makes the code harder to read and organise/maintain)

**Reply by climba03003** [reply]
`prefix` is mounted to the `instance`, when using `fastify-plugin` which means it will shared the instance with parent. Thus, `prefix` should not take any effect on it. Otherwise, the endpoint for whole `instance` will mess up.

And, you should read about the note inside document.
https://www.fastify.io/docs/latest/Reference/Plugins/#plugin-options

**Reply by climba03003** [reply]
Anyway. For `routes`, I don't see why it should wrap with `fastify-plugin`.

**Reply by hardillb** [reply]
OK, working now I've removed the `fastify-plugin` wrapper.

Adding it was just habit as I've been doing other stuff where it makes sense.

Thanks for helping me get to the bottom of this.

**Reply by jsumners** [reply]
Note, you can do:

```js
fastify.register(function (fastify){
  fastify.register(require('./routes-plugin'))
}, { prefix: '/foo' })
```

This will apply the prefix as desired.
